fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/partition/PartitionSelector.java

```
@@ -176,18 +193,46 @@ public static List<String> getPartitionNamesByExpr(ConnectContext context,
                     columnRefOperatorMap, isRecyclingCondition);
         } else if (partitionInfo.isListPartition()) {
             ListPartitionInfo listPartitionInfo = (ListPartitionInfo) partitionInfo;
-            selectedPartitionIds = getListPartitionIdsByExpr(tableName.getDb(), olapTable, listPartitionInfo,
+            selectedPartitionIds = getListPartitionIdsByExpr(tableName.getDb(), olapTable, listPartitionInfo, 
                     whereExpr, scalarOperator, exprToColumnIdxes);
         } else {
             throw new SemanticException("Unsupported partition type: " + partitionInfo.getType());
         }
         if (selectedPartitionIds == null) {
             throw new SemanticException("Failed to prune partitions with where expression: " + whereExpr.toSql());
         }
-        return selectedPartitionIds.stream()
-                .map(p -> olapTable.getPartition(p))
+        return selectedPartitionIds;
+    }
+
+    /**
+     * Get expired partitions by retention condition.
+     * inputCells and isMockPartitionIds are used for mv refresh since the partition is not added into table yet, but
+     * we need to check whether inputCells are expired or created for mv refresh.
+     */
+    public static List<String> getExpiredPartitionsByRetentionCondition(Database db,
+                                                                        OlapTable olapTable,
+                                                                        String ttlCondition) {
+        TableName tableName = new TableName(db.getFullName(), olapTable.getName());
+        ConnectContext context = ConnectContext.get() != null ? ConnectContext.get() : new ConnectContext();
+        // needs to parse the expr each schedule because it can be changed dynamically
+        // TODO: cache the parsed expr to avoid parsing it every time later.
+        Expr whereExpr = SqlParser.parseSqlToExpr(ttlCondition, SqlModeHelper.MODE_DEFAULT);
```

it would be better to report a clear error message